File: geek_java/round02/_206_reverse_Linked_List.java
package round02;

import common.ListNode;

public class _206_reverse_Linked_List {
    public static void main(String[] args) {
        ListNode node = ListNode.getSingleNode();
        ListNode.printListNode(node);
        ListNode.printListNode(reverseLinkedList(node));
    }

    private static ListNode reverseLinkedList(ListNode head) {
        if (head == null) return null;
        ListNode pre = null;
        ListNode cur = head;
        while (cur != null) {
           ListNode tmp = cur.next;
           cur.next = pre;
           pre = cur;
           cur = tmp;
        }
        return pre;
    }
}

File: src/kotlin_language/206_reverse_linked_list.kt
package kotlin_language

class `206_reverse_linked_list` {
    companion object {
        fun reverseList(head: ListNode?): ListNode? {
            if (head == null) {
                return head
            }
            var prev: ListNode? = null
            var cur = head
            while (cur != null) {
                val tmp = cur.next
                cur.next = prev
                prev = cur
                cur = tmp
            }
            return prev
        }

        fun reverseList_v1(head: ListNode?): ListNode? {
            if (head?.next == null) return head
            val tmp = reverseList_v1(head.next)
            head.next?.next = head
            head.next = null
            return tmp
        }
    }
}

fun main() {
    showListNode(`206_reverse_linked_list`.reverseList(nodeNull()))
    showListNode(`206_reverse_linked_list`.reverseList(node1()))
    showListNode(`206_reverse_linked_list`.reverseList(node2()))
    showListNode(`206_reverse_linked_list`.reverseList(node5()))
    showListNode(`206_reverse_linked_list`.reverseList_v1(nodeNull()))
    showListNode(`206_reverse_linked_list`.reverseList_v1(node1()))
    showListNode(`206_reverse_linked_list`.reverseList_v1(node2()))
    showListNode(`206_reverse_linked_list`.reverseList_v1(node5()))
}